File: 18命令模式(Command)/restaurant.py
# -*- coding:utf-8 -*-
# author: mingregister
# date：2020-04-16,19:02 

from abc import ABCMeta, abstractclassmethod


class Command(metaclass=ABCMeta):
    @abstractclassmethod
    def execute(self, *args, **kwargs):
        pass

class Restaurant():
    def __init__(self,TotalDishes, CleanedDishes):
        self.TotalDishes = TotalDishes
        self.CleanedDishes = CleanedDishes

    def MakePizza(self, n):
        return MakePizzaCommand(n, self)

    def MakeSalad(self, n):
        return MakeSaladCommand(n, self)

    def CleanDishes(self):
        return CleanDishesCommand(self)

class MakePizzaCommand(Command):
    def __init__(self, n, restaurant):
        self.n = n
        self.restaurant = restaurant

    def execute(self, *args, **kwargs):
        self.restaurant.CleanedDishes -= self.n
        # 这些print都可以改成日志记录的.
        print("made %s pizzas" % self.n)

class MakeSaladCommand(Command):
    def __init__(self, n, restaurant):
        self.n = n
        self.restaurant = restaurant

    def execute(self, *args, **kwargs):
        self.restaurant.CleanedDishes -= self.n
        print("made %d salads" % self.n)

class CleanDishesCommand(Command):
    def __init__(self, restaurant):
        self.restaurant = restaurant

    def execute(self, *args, **kwargs):
        self.restaurant.CleanedDishes = self.restaurant.TotalDishes
        print("dishes cleaned")


class Cook():
    def __init__(self, commands=None):
        # 这个逻辑也没有什么必要，但是留着吧.
        if commands is not None:
            self.commands = commands
        else:
            raise ValueError("None")

    def executeCommands(self):
        for c in self.commands:
            c.execute()

# 对于python来说，这个函数什么必要.
def NewResteraunt():
    totalDishes = 10
    CleanedDishes = totalDishes
    return Restaurant(totalDishes, CleanedDishes)

if __name__ == '__main__':
    r = NewResteraunt()
    
    # 这里是初始化了很多类的，有办法延时加载么?
    tasks = [r.MakePizza(2),
             r.MakeSalad(1),
             r.MakePizza(3),
             r.CleanDishes(),
             r.MakePizza(4),
             r.CleanDishes()
            ]


    # 不够优雅.
    tasks1 = [tasks[0], tasks[2], tasks[4]]
    tasks2 = [task for task in tasks if task not in tasks1]

    cooks = [Cook(tasks1), Cook(tasks2)]

    for i, c in enumerate(cooks):
        print("cook %d :" %i)
        c.executeCommands()

File: 18命令模式(Command)/restaurant.go
package main

// https://www.sohamkamani.com/golang/command-pattern/

import (
    "fmt"
)

// 抽象command
type Command interface {
    execute()
}

// type Command interface {
//     execute() error
// }

// The restaurant contains the total dishes and the total cleaned dishes
type Restaurant struct {
    TotalDishes   int
    CleanedDishes int
}

// `NewRestaurant` constructs a new restaurant instance with 10 dishes,
// all of them being clean
func NewResteraunt() *Restaurant {
    const totalDishes = 10
    return &Restaurant{
        TotalDishes:   totalDishes,
        CleanedDishes: totalDishes,
    }
}

// The MakePizzaCommand is a struct which contains
// the number of pizzas to make, as well as the
// restaurant as its attributes
type MakePizzaCommand struct {
    n          int
    // 传结构体，使用指针传.
    restaurant *Restaurant
}

func (c *MakePizzaCommand) execute() {
    // Reduce the total clean dishes of the restaurant
    // and print a message once done
    c.restaurant.CleanedDishes -= c.n
    fmt.Println("made", c.n, "pizzas")
}

// The MakeSaladCommand is similar to the MakePizza command
type MakeSaladCommand struct {
    n          int
    restaurant *Restaurant
}

func (c *MakeSaladCommand) execute() {
    c.restaurant.CleanedDishes -= c.n
    fmt.Println("made", c.n, "salads")
}

type CleanDishesCommand struct {
    restaurant *Restaurant
}

func (c *CleanDishesCommand) execute() {
    // Reset the cleaned dishes to the total dishes
    // present, and print a message once done
    c.restaurant.CleanedDishes = c.restaurant.TotalDishes
    fmt.Println("dishes cleaned")
}



func (r *Restaurant) MakePizza(n int) Command {
    return &MakePizzaCommand{
        restaurant: r,
        n:          n,
    }
}

func (r *Restaurant) MakeSalad(n int) Command {
    return &MakeSaladCommand{
        restaurant: r,
        n:          n,
    }
}

func (r *Restaurant) CleanDishes() Command {
    return &CleanDishesCommand{
        restaurant: r,
    }
}


// A Cook comes with their list of commands as attributes
type Cook struct {
    // 构造一个Command的slice，在初始化时传入.
    Commands []Command
}

// The executeCommands method executes all the commands
// one by one
func (c *Cook) executeCommands() {
    for _, c := range c.Commands {
        c.execute()
    }
}


func main() {
    // initialize a new resaurant
    r := NewResteraunt()

    // create the list of tasks to be executed
    tasks := []Command{
        r.MakePizza(2),
        r.MakeSalad(1),
        r.MakePizza(3),
        r.CleanDishes(),
        r.MakePizza(4),
        r.CleanDishes(),
    }

    // create the cooks that will execute the tasks
    cooks := []*Cook{
        &Cook{},
        &Cook{},
    }

    // Assign tasks to cooks alternating between the existing
    // cooks.
    for i, task := range tasks {
        // Using the modulus of the current task index, we can
        // alternate between different cooks
        cook := cooks[i%len(cooks)]
        cook.Commands = append(cook.Commands, task)
    }

    // Now that all the cooks have their commands, we can call
    // the `executeCommands` method that will have each cook
    // execute their respective commands
    for i, c := range cooks {
        fmt.Println("cook", i, ":")
        c.executeCommands()
    }
}


